store/sqlstore/retention_policy_store.go

```diff
@@ -0,0 +1,529 @@
+	if len(teamIDs) > 0 {
+		teamsSelectQuery, teamsSelectArgs, _ := s.getQueryBuilder().
+			Select("Id").
+			From("Teams").
+			Where(inStrings("Id", teamIDs)).
+			ToSql()
+		var rows []*string
+		_, err := s.GetReplica().Select(&rows, teamsSelectQuery, teamsSelectArgs...)
+		if err != nil {
+			return err
+		}
+		if len(rows) == len(teamIDs) {
+			return nil
+		}
+		retrievedIDs := make(map[string]bool)
+		for _, teamID := range rows {
+			retrievedIDs[*teamID] = true
+		}
+		for _, teamID := range teamIDs {
+			if _, ok := retrievedIDs[teamID]; !ok {
+				return store.NewErrNotFound("Team", teamID)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *SqlRetentionPolicyStore) checkChannelsExist(channelIDs []string) error {
+	if len(channelIDs) > 0 {
+		channelsSelectQuery, channelsSelectArgs, _ := s.getQueryBuilder().
+			Select("Id").
+			From("Channels").
+			Where(inStrings("Id", channelIDs)).
+			ToSql()
+		var rows []*string
+		_, err := s.GetReplica().Select(&rows, channelsSelectQuery, channelsSelectArgs...)
+		if err != nil {
+			return err
+		}
+		if len(rows) == len(channelIDs) {
+			return nil
+		}
+		retrievedIDs := make(map[string]bool)
+		for _, channelID := range rows {
+			retrievedIDs[*channelID] = true
+		}
+		for _, channelID := range channelIDs {
+			if _, ok := retrievedIDs[channelID]; !ok {
+				return store.NewErrNotFound("Channel", channelID)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *SqlRetentionPolicyStore) buildInsertRetentionPoliciesChannelsQuery(policyID string, channelIDs []string) (query string, args []interface{}) {
+	if len(channelIDs) > 0 {
+		builder := s.getQueryBuilder().
+			Insert("RetentionPoliciesChannels").
+			Columns("PolicyId", "ChannelId")
+		for _, channelID := range channelIDs {
+			builder = builder.Values(policyID, channelID)
+		}
+		query, args, _ = builder.ToSql()
+	}
+	return
+}
+
+func (s *SqlRetentionPolicyStore) buildInsertRetentionPoliciesTeamsQuery(policyID string, teamIDs []string) (query string, args []interface{}) {
+	if len(teamIDs) > 0 {
+		builder := s.getQueryBuilder().
+			Insert("RetentionPoliciesTeams").
+			Columns("PolicyId", "TeamId")
+		for _, teamID := range teamIDs {
+			builder = builder.Values(policyID, teamID)
+		}
+		query, args, _ = builder.ToSql()
+	}
+	return
+}
+
+func (s *SqlRetentionPolicyStore) Patch(patch *model.RetentionPolicyWithTeamAndChannelIds) (*model.RetentionPolicyWithTeamAndChannelCounts, error) {
+	// Strategy:
+	// 1. Update policy attributes
+	// 2. Delete existing channels from policy
+	// 3. Insert new channels into policy
+	// 4. Delete existing teams from policy
+	// 5. Insert new teams into policy
+	// 6. Read new policy
+
+	if err := s.checkTeamsExist(patch.TeamIds); err != nil {
+		return nil, err
+	}
+	if err := s.checkChannelsExist(patch.ChannelIds); err != nil {
+		return nil, err
+	}
+
+	policyUpdateQuery := ""
+	policyUpdateArgs := []interface{}{}
+	if patch.DisplayName != "" || patch.PostDuration > 0 {
```

Comment:
Do we need all of these conditionals to check if `DisplayName` and `PostDuration` are not their default values? Or can we just invoke `builder.Set` regardless?

Reply:
We could, but it saves us from running an unnecessary UPDATE statement.